File: tutorials/testing/java/tut_sllist_p3.java
import bridges.connect.Bridges;
import bridges.base.SLelement;
import java.util.Iterator;

//
// In part 3 of this tutorial we demonstrate more advanced features such as
// using iterators to traverse lists, which are more convenient than working
// with the underlying element objects. To keep the tutorial simple, we will ignore
// styling (that was illustrated in part 2).
//
// Reference Classes: SLelement, Bridges
//


public class tut_sllist_p3 {

	public static void main(String[] args) throws Exception {

		// Create Bridges object,  fill in credentials
		Bridges bridges = new Bridges(YOUR_ASSIGNMENT_NUMBER, "YOUR_USER_ID",
			"YOUR_API_KEY");
		// set title
		bridges.setTitle("A Single Linked List Example");

		// set  description
		bridges.setDescription("Demonstrate advanced features, looping, iterators");

		// create the linked list elements with student names
		SLelement<String>  st0 = new SLelement<String> ("Gretel Chaney");
		SLelement<String>  st1 = new SLelement<String> ("Lamont Kyler");
		SLelement<String>  st2 = new SLelement<String> ("Gladys Serino");
		SLelement<String>  st3 = new SLelement<String> ("Karol Soderman");
		SLelement<String>  st4 = new SLelement<String> ("Starr McGinn");

		//  link the elements
		st0.setNext(st1);
		st1.setNext(st2);
		st2.setNext(st3);
		st3.setNext(st4);

		// we want to see these names in the visualization so we will set them as
		// the nodes' labels. We will retrieve the nodes' generic data for the label
		st0.setLabel(st0.getValue());
		st1.setLabel(st1.getValue());
		st2.setLabel(st2.getValue());
		st3.setLabel(st3.getValue());
		st4.setLabel(st4.getValue());

		// print the names of the students in the list using iteration

		// Using a regular for loop on BRIDGES elements
		System.out.println("Using a regular for loop..");
		for (SLelement<String> el = st0; el != null; el = el.getNext())
			System.out.println("\t" +  el.getValue());

		// using a forward iterator
		System.out.println("Using a forward iterator..\n");
		Iterator<String> iter = st0.iterator();
		while (iter.hasNext()) {
			System.out.println("\t" + iter.next());
		}

		// using a for-each loop
		System.out.println("Using range loop:");
		for (String s : st0)
			System.out.println("\t " + s);

		// tell Bridges the head of the list
		bridges.setDataStructure(st0);

		// visualize the linked list (hit the key 'l' on the visualzation to see all labels
		bridges.visualize();
	}
}

File: tutorials/testing/python/tut_sllist_p3.py
from bridges.bridges import *
from bridges.sl_element import *
import sys

#
# In part 3 of this tutorial we demonstrate more advanced features such as 
# using iterators to traverse lists, which are more convenient than working
# with the underlying element objects. To keep the tutorial simple, we will ignore
# styling (that was illustrated in part 2).
#
# Reference Classes: SLelement, Bridges 
#


def main():

    # create the Bridges object, set credentials
    bridges = Bridges(YOUR_ASSSIGNMENT_NUMBER, "YOUR_USER_ID", "YOUR_API_KEY")

    # set title, description
    bridges.set_title("A Singly Linked List Example")
    bridges.set_description("Demonstrate advanced features, looping, iterators")

	# create the list elements
    st0 = SLelement(e="Gretel Chaney")
    st1 = SLelement(e="Lamont Kyler")
    st2 = SLelement(e="Gladys Serino")
    st3 = SLelement(e="Karol Soderman")
    st4 = SLelement(e="Starr McGinn")

    # link the elements
    st0.next = st1
    st1.next = st2
    st2.next = st3
    st3.next = st4

    # we want to see these names in the visualization so we will 
    #set them as the nodes' labels. We will retrieve the nodes' 
    #generic data for the label
    st0.label = st0.value
    st1.label = st1.value
    st2.label = st2.value
    st3.label = st3.value
    st4.label = st4.value

    # print the names of the students in the list using iteration

    # Using a regular for loop on BRIDGES elements
    print("\nUsing a regular while loop..\n"); 
    el = st0
    while el != None:
        print (el.value)
        el = el.next


    print("\nUsing a list iterator..\n"); 
    list_iter = SLelementIterator(st0);
    i = 0
    st = st0
    while list_iter.has_next():
        print (st.value)
        st = list_iter.next()
        i = i+1

    # tell Bridges the head of the list
    bridges.set_data_structure(st0)

    # visualize
    bridges.visualize()


if __name__ == "__main__":
    main()

